kernel/src/misc/wm/wm.c

```diff
@@ -489,65 +509,55 @@ void wm_mouse_callback(mouse_t raw_curr) {
     if (!raw_prev.left_pressed && raw_curr.left_pressed) {
-        wm_window_t* clicked = wm_window_at(mouse.x, mouse.y);
+        lmb_is_down = true;
+    } else if (!raw_curr.left_pressed) {
+        lmb_is_down = false;
+    }
 
-        if (clicked) {
-            wm_raise_window(clicked);
-            dragged = clicked;
+    // Raise window
+    if (lmb_is_down){
+        if ( (win = wm_window_at(mouse.x, mouse.y)) ) {
+            wm_raise_window(win);
         }
     }
 
-    // Drags
-    if (raw_prev.left_pressed && raw_curr.left_pressed) {
-        if (dragged && (dx || dy)) {
-            rect_t rect = rect_from_window(dragged);
+    // Drag window
+    if (lmb_is_down) {
+        // only drag window when hovering title bar and 
+        if ((dx || dy) && wm_hovering_tb_at(mouse.x, mouse.y)) {
+            rect_t rect = rect_from_window(win);
 
-            if (!(rect.left + dx < 0 || rect.right + dx >= (int32_t) fb.width ||
-                    rect.top + dy < 0 || rect.bottom + dy >= (int32_t) fb.height)) {
-                been_dragged = true;
+            win->x += dx;
+            win->y += dy;
 
-                dragged->x += dx;
-                dragged->y += dy;
+            rect_t new_rect = rect_from_window(win);
 
-                rect_t new_rect = rect_from_window(dragged);
+            wm_refresh_partial(rect);
+            wm_draw_window(win, new_rect);
 
-                wm_refresh_partial(rect);
-                wm_draw_window(dragged, new_rect);
-
-                rect = wm_mouse_to_rect(mouse);
-                wm_draw_mouse(rect);
-            }
+            rect = wm_mouse_to_rect(mouse);
+            wm_draw_mouse(rect);
         }
-    }
-
     // Clicks and drag ends
-    if (raw_prev.left_pressed && !raw_curr.left_pressed) {
+    } else if (raw_prev.left_pressed) {
         wm_event_t event;
+        rect_t r = rect_from_window(win);
 
-        if (!been_dragged) {
-            rect_t r = rect_from_window(dragged);
-
-            event.type = WM_EVENT_CLICK;
-            event.mouse.position = wm_mouse_to_rect(mouse);
-            event.mouse.position.top -= r.top;
-            event.mouse.position.left -= r.left;
-
-            ringbuffer_write(dragged->events, sizeof(wm_event_t), (uint8_t*) &event);
-        }
+        event.type = WM_EVENT_CLICK;
+        event.mouse.position = wm_mouse_to_rect(mouse);
+        event.mouse.position.top -= r.top;
+        event.mouse.position.left -= r.left;
 
-        been_dragged = false;
-        dragged = NULL;
-    }
+        ringbuffer_write(win->events, sizeof(wm_event_t), (uint8_t*) &event);
 
-    // Simple moves
-    if (!raw_prev.left_pressed && !raw_curr.left_pressed) {
+    } else {
         wm_window_t* under_cursor = wm_window_at(mouse.x, mouse.y);
-        wm_event_t event;
 
         if (under_cursor) {
-            rect_t r = rect_from_window(under_cursor);
+            wm_event_t event;
+            rect_t r = rect_from_window(win);
```

Comment:
The crash is here, when this function is first called because of a mouse move: `win` is null. There are other branches in this function where `win` might be null that will cause problems.

Reply:
using `under_cursor` makes it not crash.

Reply:
thanks. I'll keep it up from here